llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp

```diff
@@ -605,9 +605,12 @@ createScalarIVSteps(VPlan &Plan, InductionDescriptor::InductionKind Kind,
                     VPBuilder &Builder) {
   VPRegionBlock *LoopRegion = Plan.getVectorLoopRegion();
   VPBasicBlock *HeaderVPBB = LoopRegion->getEntryBasicBlock();
-  VPCanonicalIVPHIRecipe *CanonicalIV = LoopRegion->getCanonicalIV();
-  VPSingleDefRecipe *BaseIV = Builder.createDerivedIV(
-      Kind, FPBinOp, StartV, CanonicalIV, Step, "offset.idx");
+  VPHeaderPHIRecipe *IV = LoopRegion->getCanonicalIV();
+  if (auto *EVLIV =
+          dyn_cast<VPEVLBasedIVPHIRecipe>(std::next(IV->getIterator())))
```

Comment:
Do you think we can avoid this by deferring canonical IV replacement until canonicalizeEVLLoops?

Reply:
I don't think thats a good idea because it means we will have an incorrect VPlan throughout the optimisation pipeline. 

Part of the motivation for this PR is to have everything use the EVL based IV as soon as it's added so we don't accidentally have recipes using the canonical IV and producing incorrect results on the penultimate iteration.

We could probably add a method to VPRegionBlock that abstracts over the EVL or canonical IV like `getEffectiveIV`, but that probably requires more discussion so I'd like to leave that to another PR if possible